### PR TITLE
Security: remove put privilege API

### PR DESCRIPTION
**Pull request by jaymode**
This commit removes the put privilege API in favor of having a single API to
create and update privileges. If we see the need to have an API like this in
the future we can always add it back.

### CODE REVIEW

File: x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/10_basic.yml [hunk 1]
```diff
@@ -30,24 +30,26 @@ teardown:
...
+      xpack.security.put_privileges:
         body:  >
             {
-              "application": "app",
-              "name": "p1",
-              "actions": [ "data:read/*" , "action:login" ],
-              "metadata": {
-                "key1" : "val1a",
-                "key2" : "val2a"
+              "app": {
```
Comment (Mpdreamz) [inline x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/10_basic.yml]: Whats the relation between `app` and `p1` and them being later declared again using `application: "app"` and `name: "p1"` do these need to be equal?

If we are taking a `Map<AppName, Map<PrivilegeName, Privilege>>` I think we should omit `application` and `name` or we have to take `Array<Privilege>`

Comment (jaymode) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/10_basic.yml]: They are equivalent as far as I know and I agree that it is redundant. @tvernum can you chime in about this aspect?

Comment (tvernum) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/10_basic.yml]: They're redundant and optional, but if they exist they need to match the keys from the containing object.
They're allowed for three reasons
1. Historical, originally I had the body just being an array, but there are a few problems with array bodies, including that rest tests can't handle it, so I switched to objects, and at Albert's suggestion made the inner fields optional.
2. It means that the objects that come from the GET API can be fed back into the PUT/POST without needing to remove fields
3. The format for index storage needs the fields and we re-use the parser, so we needed to do something and optional seemed better than rejection.

**Comment by tvernum** [comment]
This looks good but I'm worried that the semantics aren't quite correct for a PUT (See review comment with link to previous discussion)

File: x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/10_basic.yml (hunk 1 above)
Comment (jaymode) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/privileges/10_basic.yml]: Optional makes sense. I updated the rest test to use it both ways

**Comment by jaymode** [comment]
@Mpdreamz we discussed this in our team meeting. We came to the conclusion that this will be a post only API since it does not follow the PUT semantics. Given that, we feel the name of the spec is misleading but didn't have a great name for it. This API will create or update a given privilege. Do you have any naming suggestions? IIUC this name might be important for the generation of the client from the spec files so we want to make sure its right for you guys.

**Comment by jaymode** [comment]
@tvernum can you take a look and verify everything looks good other than any potential naming change that we make

**Comment by Mpdreamz** [comment]
@tvernum mind linking to the discussion (do not spot it). I'm actually OK with the `put_` prefix and having `["PUT", "POST"]` as its not that different from: [cluster.put_settings.json](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/cluster.put_settings.json), [indices.put_settings.json](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/indices.put_settings.json) & [indices.put_mapping.json](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/indices.put_mapping.json). 

>This API will create or update a given privilege. 

PUT allows for this but typically means it is replacing the resource completely if it exists already. The three linked existing API's do not follow that rule but do follow PUT's idempotent requirement which I think `put_privileges.json` does as well. 

> It means that the objects that come from the GET API can be fed back into the PUT/POST without needing to remove fields

Ideally the `GET` for privileges also returns a `Map<AppName, Map<PrivilegeName, Privilege>>` thats what e.g [indices.get_template.json](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/indices.get_template.json) & [indices.get_alias.json](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/indices.get_alias.json) and many others do as well without repeating the two key names inside the nested definition.

**Comment by jaymode** [comment]
> Ideally the GET for privileges also returns a Map<AppName, Map<PrivilegeName, Privilege>> thats what e.g indices.get_template.json & indices.get_alias.json and many others do as well without repeating the two key names inside the nested definition.

It does return the Map but does repeat the two keys inside the definition

**Comment by jaymode** [comment]
I chatted with @kobelb and they are currently using the full response. Given the timing, my plan is to leave this as is and for 7.0 we can remove the duplication in the response.

**Comment by Mpdreamz** [comment]
@jaymode fair enough, thank you and @tvernum for the quick turnaround on this! 👍

**Review by Mpdreamz — APPROVED** [review]

**Comment by tvernum** [comment]
Sorry @Mpdreamz the discussion was https://github.com/elastic/elasticsearch/issues/31730
I didn't realise my comments was pending (github mobile is simply terrible)

**Review by tvernum — APPROVED** [review]
👍

**Pull request merged**